File: Algorithm/doublePoint.php
<?php
function dd($val){
    var_dump($val);
    exit();
}

function maxArea($arr) {
    $max = 0;
    [$i ,$j] = [0,count($arr)-1];
    while($i < $j){
        $max = max($max,(min($arr[$i],$arr[$j]))*($j -$i));
        $arr[$i] > $arr[$j] ? $j-- : $i++;
    }
    return $max;
}

function pivotIndex($nums) {
    //想复杂了，直接遍历两次就好了
    //伪代码
    $sum = array_sum($nums);
    [$index, $i] = [0, 0];
    $j = $sum;
    
    $arr = array();
    
    for($index;$index < count($nums); $index++ ){
        $j -= $nums[$index];
        if($i == $j){
            $arr[] = $index;
        }
        $i += $nums[$index]; 
    }
    
    if(count($arr) > 0) return min($arr);
    return -1;
}

function isSubsequence($s, $t) {
    if(empty($s)) return true;
    $index = 0;
    for($i=0;$i<strlen($t);$i++){
        if($t[$i] == $s[$index]){
            $index++;
            if(strlen($s) == $index) return true;
        }
    }
    return false;
}

function isIsomorphic($s, $t) {
    //先找思路，再做题
    //先遍历这个字符串
    //把值作为键名，把键名作为值存储为数组
    //再遍历后面一个
    //比较他们两个的值。所有相等则相等
    if(strlen($s) != strlen($t)) return false;
    $pattern = [];
    $object = [];
    for($i=0;$i<strlen($s);$i++){
        $pattern[$s[$i]][] = $i;
        $object[$t[$i]][] = $i;
        if($pattern[[$s[$i]]] != $object[$t[$i]]){
            return false;
        }
    }
    
    return true;
}

function mergeTwoLists($l1, $l2){
    //边界控制
    if($l1->next == null) return $l1;
    if($l2->next == null) return $l2;

    if($l1->val <= $l2->val){
        $l1->next = mergeTwoLists($l1->next, $l2);
        return $l1;
    }else{
        $l2->next = mergeTwoLists($l1, $l2->next);
        return $l2;
    }

}

function main(){
//    $array = [12, 2, 3, 124, 99, 0, 234, 34, 11, 345, 51];
    // dd(maxArea($array));
    // dd(pivotIndex([1, 7, 3, 6, 5, 6]));
    // dd(isIsomorphic($a = 'abd', $b = 'bac'));
    $a =(object)[
        'val'=>1,
        'next' => (object)[
            'val'=>2,
            'next' => (object)[
                'val'=>3,
                'next' => (object)[
                    'val'=>4,
                    'next' => NULL
                ]
            ]
        ]
    ];
    $b =(object)[
        'val'=>1,
        'next' => (object)[
            'val'=>2,
            'next' => (object)[
                'val'=>3,
                'next' => (object)[
                    'val'=>7,
                    'next' => NULL
                ]
            ]
        ]
    ];

dd(mergeTwoLists($a,$b));
}

main();

File: Algorithm/doublePoint.go
package main

import (
	"fmt"
	"math"
)

func main() {
	//pivotIndex([]int{1, 7, 3, 6, 5, 6})
	//testS([]int{1, 2, 3, 4, 5})
	isSubsequence("hello", "world")
}

func maxArea(arr []int) int {
	max := 0
	i, j := 0, len(arr)-1
	for i < j {
		min := int(math.Min(float64(arr[i]), float64(arr[j])))
		max = int(math.Max(float64(max), float64(min*(j-i))))
		if arr[i] > arr[j] {
			j--
		} else {
			i++
		}
	}
	return max
}

func pivotIndex(nums []int) int {
	var sum int
	var arr []int
	i, index := 0, 0
	for _, v := range nums {
		sum += v
	}
	j := sum
	for index = 0; index < len(nums); index++ {
		j -= nums[index]
		if i == j {
			arr = append(arr, index)
		}
		i += nums[index]
	}
	if len(arr) > 0 {
		return arr[0]
	}
	return -1
}

func convArray(arr []int) []int {
	for i, j := 0, len(arr)-1; i < j; i, j = i+1, j-1 {
		arr[i], arr[j] = arr[j], arr[i]
	}
	return arr
}

// 测试slice会不会因为array的改变而改变
func testS(arr []int) {
	a := arr[:]
	for k, v := range arr {
		arr[k] = v * v
	}
	fmt.Print(arr)
	fmt.Print("\r\n")
	fmt.Print(a)
}

func isSubsequence(h string, n string) bool {
	//思路一定要清晰
	//1.首先应该确定遍历哪个 n
	//2.先拿出来 h 中的 一个，然后遍历 n ，如果 n存在，则 h中的指针往后移动一步，n继续跑
	index := 0
	for k, _ := range n {
		if h[index] == n[k] {
			index++
			if index == len(h) {
				return true
			}
		}
	}
	return false
}
